src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -51,39 +56,497 @@ internal static class Utf8Encoder
+                    {
+                        ch = *pSrc;
+                        pSrc++;
+
+                        if (ch > 0x7F)
+                        {
+                            goto LongCode;
+                        }
+                        *pTarget = (char)ch;
+                        pTarget++;
+
+                        // get pSrc to be 2-byte aligned
+                        if ((unchecked((int)pSrc) & 0x1) != 0)
+                        {
+                            ch = *pSrc;
+                            pSrc++;
+                            if (ch > 0x7F)
+                            {
+                                goto LongCode;
+                            }
+                            *pTarget = (char)ch;
+                            pTarget++;
+                        }
+
+                        // get pSrc to be 4-byte aligned
+                        if ((unchecked((int)pSrc) & 0x2) != 0)
+                        {
+                            ch = *(ushort*)pSrc;
+                            if ((ch & 0x8080) != 0)
+                            {
+                                goto LongCodeWithMask16;
+                            }
+
+                            // Unfortunately, this is endianess sensitive
+    #if BIGENDIAN
+                            *pTarget = (char)((ch >> 8) & 0x7F);
+                            pSrc += 2;
+                            *(pTarget+1) = (char)(ch & 0x7F);
+                            pTarget += 2;
+    #else // BIGENDIAN
+                            *pTarget = (char)(ch & 0x7F);
+                            pSrc += 2;
+                            *(pTarget + 1) = (char)((ch >> 8) & 0x7F);
+                            pTarget += 2;
+    #endif // BIGENDIAN
+                        }
+
+                        // Run 8 characters at a time!
+                        while (pTarget < pStop)
+                        {
+                            ch = *(int*)pSrc;
+                            int chb = *(int*)(pSrc + 4);
+                            if (((ch | chb) & unchecked((int)0x80808080)) != 0)
+                            {
+                                goto LongCodeWithMask32;
+                            }
+
+                            // Unfortunately, this is endianess sensitive
+    #if BIGENDIAN
+                            *pTarget = (char)((ch >> 24) & 0x7F);
+                            *(pTarget+1) = (char)((ch >> 16) & 0x7F);
+                            *(pTarget+2) = (char)((ch >> 8) & 0x7F);
+                            *(pTarget+3) = (char)(ch & 0x7F);
+                            pSrc += 8;
+                            *(pTarget+4) = (char)((chb >> 24) & 0x7F);
+                            *(pTarget+5) = (char)((chb >> 16) & 0x7F);
+                            *(pTarget+6) = (char)((chb >> 8) & 0x7F);
+                            *(pTarget+7) = (char)(chb & 0x7F);
+                            pTarget += 8;
+    #else // BIGENDIAN
```

Comment:
This ifdef is for endianness of the underlying platform, not for endianness of the encoding...